specs/capella/beacon-chain.md

```diff
@@ -0,0 +1,305 @@
+    # Slashings
+    slashings: Vector[Gwei, EPOCHS_PER_SLASHINGS_VECTOR]  # Per-epoch sums of slashed effective balances
+    # Participation
+    previous_epoch_participation: List[ParticipationFlags, VALIDATOR_REGISTRY_LIMIT]
+    current_epoch_participation: List[ParticipationFlags, VALIDATOR_REGISTRY_LIMIT]
+    # Finality
+    justification_bits: Bitvector[JUSTIFICATION_BITS_LENGTH]  # Bit set for every recent justified epoch
+    previous_justified_checkpoint: Checkpoint
+    current_justified_checkpoint: Checkpoint
+    finalized_checkpoint: Checkpoint
+    # Inactivity
+    inactivity_scores: List[uint64, VALIDATOR_REGISTRY_LIMIT]
+    # Sync
+    current_sync_committee: SyncCommittee
+    next_sync_committee: SyncCommittee
+    # Execution
+    latest_execution_payload_header: ExecutionPayloadHeader
+    # Withdrawals
+    withdrawal_receipts: List[WithdrawalTransaction, WITHDRAWAL_TRANSACTION_LIMIT]  # [New in Capella]
+```
+
+#### `ExecutionPayload`
+
+```python
+class ExecutionPayload(Container):
+    # Execution block header fields
+    parent_hash: Hash32
+    fee_recipient: ExecutionAddress  # 'beneficiary' in the yellow paper
+    state_root: Bytes32
+    receipts_root: Bytes32
+    logs_bloom: ByteVector[BYTES_PER_LOGS_BLOOM]
+    prev_randao: Bytes32  # 'difficulty' in the yellow paper
+    block_number: uint64  # 'number' in the yellow paper
+    gas_limit: uint64
+    gas_used: uint64
+    timestamp: uint64
+    extra_data: ByteList[MAX_EXTRA_DATA_BYTES]
+    base_fee_per_gas: uint256
+    # Extra payload fields
+    block_hash: Hash32  # Hash of execution block
+    transactions: List[Transaction, MAX_TRANSACTIONS_PER_PAYLOAD]
+    withdrawal_transactions: List[WithdrawalTransaction, MAX_WITHDRAWAL_TRANSACTIONS_PER_PAYLOAD]  # [New in Capella]
+```
+
+#### `ExecutionPayloadHeader`
+
+```python
+class ExecutionPayloadHeader(Container):
+    # Execution block header fields
+    parent_hash: Hash32
+    fee_recipient: ExecutionAddress
+    state_root: Bytes32
+    receipts_root: Bytes32
+    logs_bloom: ByteVector[BYTES_PER_LOGS_BLOOM]
+    prev_randao: Bytes32
+    block_number: uint64
+    gas_limit: uint64
+    gas_used: uint64
+    timestamp: uint64
+    extra_data: ByteList[MAX_EXTRA_DATA_BYTES]
+    base_fee_per_gas: uint256
+    # Extra payload fields
+    block_hash: Hash32  # Hash of execution block
+    transactions_root: Root
+    withdrawal_transactions_root: Root  # [New in Capella]
+```
+
+### New containers
+
+#### `WithdrawalTransaction`
+
+New EIP-2718 transaction type, with the format being the single byte `TX_TYPE_WITHDRAWAL`
+followed by an SSZ encoding of the `WithdrawalTransaction` container comprising the transaction contents.
+
+*Note*: This container is used for both a special TX that goes into an `ExecutionPayload`
+as well as in the `BeaconState`'s `withdrawal_receipts` queue.
+
+```python
+class WithdrawalTransaction(Container):
+    address: ExecutionAddress
+    amount: Gwei
```

Comment:
Is the amount defined in gwei to save on serialization size?

Reply:
we don't currently have uint256 arithmetic in CL so avoiding the conversion to a larger bitsize in CL currently. That said, I'm not sold on that. It can and maybe should be converted here so it is a native EL value before landing in there

Reply:
i can say while working here: https://github.com/ethereum/go-ethereum/pull/24468

it would simplify a lot of EL implementation if this was a `Bytes32`  value -- right now, accepting Gwei means we have to convert from Gwei to Wei at every IO boundary into the EL and implementers have to remember the conversion needs to be done.

it does seem simpler to do something like this once and then drop all of the conversion inside/around the EL:


```python
amount_in_gwei = ssz.encode(amount)
amount_for_receipt = amount_in_gwei + "\x00" * 24
```

Reply:
I don't think its ideal to have the value field be in different units depending on the transaction type from the pov of the EL. In its current form, would `Gwei` be expected to be a uint64? I don't see a `Gwei` typedef. Adding uint256 arithmetic to the CL feels a bit leaky, but it could be safer to use to represent ether balances as to not make assumptions about the ether supply far into the future

Reply:
yeah, I agree. It was more of a placeholder with an existing type before doing the work to add Wei here